README.md
# pino-papertrail

[![CI](https://github.com/ovhemert/pino-papertrail/workflows/CI/badge.svg)](https://github.com/ovhemert/pino-papertrail/actions)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/8c926c3289f94f1fae220df626ebd9db)](https://www.codacy.com/app/ovhemert/pino-papertrail?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=ovhemert/pino-papertrail&amp;utm_campaign=Badge_Grade)
[![Known Vulnerabilities](https://snyk.io/test/npm/pino-papertrail/badge.svg)](https://snyk.io/test/npm/pino-papertrail)
[![Coverage Status](https://coveralls.io/repos/github/ovhemert/pino-papertrail/badge.svg?branch=master)](https://coveralls.io/github/ovhemert/pino-papertrail?branch=master)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](http://standardjs.com/)

This module provides a "transport" for [pino][pino] that forwards
messages to the [papertrail][papertrail] log service through udp/tcp/tls. The module can echo the received logs or work silently.

## Installation

To use globally from command line:

```bash
npm install -g pino-papertrail
```

To include as a library in your project:

```bash
npm install pino-papertrail --save
```

## CLI

Want to use `pino-papertrail` from the CLI? See the [CLI](./docs/CLI.md) documentation for details.

## API

Want to use `pino-papertrail` as a library in your project? See the [API](./docs/API.md) documentation for details.

## Maintainers

Osmond van Hemert
[![Github](https://img.shields.io/badge/-website.svg?style=social&logoColor=333&logo=github)](https://github.com/ovhemert)
[![Web](https://img.shields.io/badge/-website.svg?style=social&logoColor=333&logo=nextdoor)](https://ovhemert.dev)

## Contributing

If you would like to help out with some code, check the [details](./docs/CONTRIBUTING.md).

Not a coder, but still want to support? Have a look at the options available to [donate](https://ovhemert.dev/donate).

## License

Licensed under [MIT](./LICENSE).

[pino]: https://www.npmjs.com/package/pino
[papertrail]: https://papertrailapp.com


docs/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Changes are grouped by:
`Added` for new features
`Changed` for changes in existing functionality
`Deprecated` for soon-to-be removed features
`Removed` for now removed features
`Fixed` for any bug fixes
`Security` in case of vulnerabilities

## [Unreleased](https://github.com/ovhemert/pino-papertrail/compare/v2.1.0...HEAD)

...

## [2.1.0](https://github.com/ovhemert/pino-papertrail/compare/v2.0.0...v2.1.0) - 2020-07-28

### Added

- Support and testing for Node.js v14.x by [@ovhemert](https://github.com/ovhemert)

### Security

- Fixed external vulnerabilities by upgrading dependencies by [@ovhemert](https://github.com/ovhemert)

## [2.0.0](https://github.com/ovhemert/pino-papertrail/compare/v1.3.0...v2.0.0) - 2020-06-02

### Added

- Backoff strategy for any tls/tcp socket errors by [@andyli](https://github.com/andyli)
- Documentation for code quality by [@ovhemert](https://github.com/ovhemert)
- TypeScript types by [@jaulz](https://github.com/jaulz)

### Fixed

- Resume logging after UDP failure [@mattiash](https://github.com/mattiash)

### Removed

- Node v8.x support by [@ovhemert](https://github.com/ovhemert)

## [1.3.0](https://github.com/ovhemert/pino-papertrail/compare/v1.2.5...v1.3.0) - 2019-09-16

### Added

- TCP and TLS support by [@andyli](https://github.com/andyli)

## [1.2.5](https://github.com/ovhemert/pino-papertrail/compare/v1.2.0...v1.2.5) - 2019-08-04

### Added

- Support and testing for Node.js v12 by [@ovhemert](https://github.com/ovhemert)

## [1.2.0](https://github.com/ovhemert/pino-papertrail/compare/v1.1.0...v1.2.0) - 2019-03-28

### Added

- API with associated tests and docs by [@kjng](https://github.com/kjng)

### Security

- Update dependencies to prevent vulnerability issues by [@ovhemert](https://github.com/ovhemert)

## [1.1.0](https://github.com/ovhemert/pino-papertrail/compare/v1.0.4...v1.1.0) - 2019-01-04

### Added

- A --message-only parameter

### Fixed

- Port number argument for testing by [@ovhemert](https://github.com/ovhemert)
- Test with hardcoded year

## [1.0.4](https://github.com/ovhemert/pino-papertrail/compare/v1.0.1...v1.0.4) - 2018-09-18

### Added

- More tests to get 100% coverage by [@ovhemert](https://github.com/ovhemert)
- Testing on latest pino version by [@ovhemert](https://github.com/ovhemert)

### Fixed

- Missing `through2` module after dependency upgrades by [@ovhemert](https://github.com/ovhemert)

## [1.0.1](https://github.com/ovhemert/pino-papertrail/releases/tag/v1.0.1) - 2018-07-20

- Initial version by [@ovhemert](https://github.com/ovhemert)


test/fixtures/logit.js
'use strict'

const pino = require('pino')
const log = pino({ level: 'trace' })

log.fatal('fatal message')
log.debug('debug message')
log.info('info message')
log.warn('warning message')
log.error(new Error('error message'))


test/fixtures/messages.js
'use strict'

module.exports = {
  traceMessage: '{"level":10,"time":1532081790741,"msg":"trace message","pid":9118,"hostname":"Osmonds-MacBook-Pro.local","v":1}',
  debugMessage: '{"level":20,"time":1532081790742,"msg":"debug message","pid":9118,"hostname":"Osmonds-MacBook-Pro.local","v":1}',
  infoMessage: '{"level":30,"time":1532081790743,"msg":"info message","pid":9118,"hostname":"Osmonds-MacBook-Pro.local","v":1}',
  warnMessage: '{"level":40,"time":1532081790744,"msg":"warning message","pid":9118,"hostname":"Osmonds-MacBook-Pro.local","v":1}',
  errorMessage: '{"level":50,"time":1532081790745,"msg":"error message","pid":9118,"hostname":"Osmonds-MacBook-Pro.local","type":"Error","stack":"Error: error message","v":1}',
  fatalMessage: '{"level":60,"time":1532081790746,"msg":"fatal message","pid":9118,"hostname":"Osmonds-MacBook-Pro.local","v":1}'
}


test/pino-papertrail.test.js
'use strict'

const dgram = require('dgram')
const path = require('path')
const os = require('os')
const spawn = require('child_process').spawn
const test = require('tap').test
const sinon = require('sinon')
const net = require('net')
const tls = require('tls')

const messages = require(path.join(__dirname, 'fixtures', 'messages'))
const appPath = path.join(path.resolve(__dirname, '..', 'cli'))
const api = require(path.join(__dirname, '..', 'index'))

test('displays help', (t) => {
  t.plan(1)
  const app = spawn('node', [appPath, '-h'])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    const res = (msg.indexOf('display help') > 0)
    t.ok(res)
  })
})

test('displays version info', (t) => {
  t.plan(1)
  const app = spawn('node', [appPath, '-v'])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    const res = msg.startsWith('pino-papertrail v')
    t.ok(res)
  })
})

test('skips non-json input', (t) => {
  t.plan(1)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    t.fail('should not receive any data')
  })
  app.on('close', (code) => {
    t.is(code, 0)
  })
  app.stdin.end('this is not json\n')
})

test('parses message without time-prop', (t) => {
  t.plan(3)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<14>1 ' + (new Date()).getFullYear()))
    t.ok(msg.includes('"level":30'))
    t.ok(msg.includes('"info message"'))
    app.kill()
  })
  const msg = messages.infoMessage.replace('"time":1532081790743,', '')
  app.stdin.end(`${msg}\n`)
})

test('parses trace message', (t) => {
  t.plan(3)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<15>1 2018'))
    t.ok(msg.includes('"level":10'))
    t.ok(msg.includes('"trace message"'))
    app.kill()
  })
  app.stdin.end(`${messages.traceMessage}\n`)
})

test('parses debug message', (t) => {
  t.plan(3)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<15>1 2018'))
    t.ok(msg.includes('"level":20'))
    t.ok(msg.includes('"debug message"'))
    app.kill()
  })
  app.stdin.end(`${messages.debugMessage}\n`)
})

test('parses info message', (t) => {
  t.plan(3)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<14>1 2018'))
    t.ok(msg.includes('"level":30'))
    t.ok(msg.includes('"info message"'))
    app.kill()
  })
  app.stdin.end(`${messages.infoMessage}\n`)
})

test('parses warn message', (t) => {
  t.plan(3)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<12>1 2018'))
    t.ok(msg.includes('"level":40'))
    t.ok(msg.includes('"warning message"'))
    app.kill()
  })
  app.stdin.end(`${messages.warnMessage}\n`)
})

test('parses error message', (t) => {
  t.plan(3)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<11>1 2018'))
    t.ok(msg.includes('"level":50'))
    t.ok(msg.includes('"error message"'))
    app.kill()
  })
  app.stdin.end(`${messages.errorMessage}\n`)
})

test('parses fatal message', (t) => {
  t.plan(3)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<10>1 2018'))
    t.ok(msg.includes('"level":60'))
    t.ok(msg.includes('"fatal message"'))
    app.kill()
  })
  app.stdin.end(`${messages.fatalMessage}\n`)
})

test('parses fatal message', (t) => {
  t.plan(3)
  const app = spawn('node', [appPath])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<10>1 2018'))
    t.ok(msg.includes('"level":60'))
    t.ok(msg.includes('"fatal message"'))
    app.kill()
  })
  app.stdin.end(`${messages.fatalMessage}\n`)
})

test('can forward only message to papertrail', (t) => {
  t.plan(4)
  const app = spawn('node', [appPath, '-m'])
  app.stdout.on('data', (data) => {
    const msg = data.toString()
    t.ok(msg.startsWith('<10>1 2018'))
    t.ok(!msg.includes('"level":60'))
    t.ok(!msg.includes('"fatal message"'))
    t.ok(msg.includes('fatal message'))
    app.kill()
  })
  app.stdin.end(`${messages.fatalMessage}\n`)
})

test('sends to udp', (t) => {
  t.plan(2)
  const app = spawn('node', [appPath, '-e', true, '-p', '12345'])
  app.stdout.on('data', (data) => {
    t.fail('should not receive any data')
  })
  app.on('close', (code) => {
    socket.close()

    if (os.type() === 'Windows_NT') {
      t.is(code, null)
    } else {
      t.is(code, 0)
    }
  })
  const socket = dgram.createSocket('udp4')
  socket.on('message', msg => {
    const res = msg.toString()
    t.ok(res.startsWith('<14>1 2018'))
    app.kill()
  })
  socket.on('error', () => {
    t.fail('socket error')
    app.kill()
  })
  socket.bind({ address: '127.0.0.1', port: 12345 }, (err) => {
    if (err) {
      t.fail('socket bind error')
      return app.kill()
    }
    app.stdin.end(`${messages.infoMessage}\n`)
  })
})

test('logs error when DNS resolution fails', (t) => {
  t.plan(2)
  const app = spawn('node', [appPath, '-H', 'nosuchhost.example.com', '-e', true, '-p', '12345'])
  app.stdout.on('data', (data) => {
    t.ok(data.toString().startsWith('error'), 'shall log an error')
    app.kill()
  })
  app.on('close', (code) => {
    if (os.type() === 'Windows_NT') {
      t.is(code, null)
    } else {
      t.is(code, 0)
    }
  })
  app.stdin.end(`${messages.infoMessage}\n`)
})

test('pino-papertrail api (no option)', (t) => {
  t.plan(1)
  t.ok(api.createWriteStream(), 'should be able to pass no options')
})

test('pino-papertrail api (udp)', (t) => {
  t.plan(1)

  const options = {
    appname: 'pino-papertrail',
    echo: false,
    host: 'papertrailapp.com',
    port: '1234',
    connection: 'udp',
    'message-only': false
  }

  t.ok(api.createWriteStream(options), 'should be able to pass options')
})

function testApiTcp (t, connection, echo) {
  t.plan(11)

  const clock = require('@sinonjs/fake-timers').install()

  const socket = {
    setKeepAlive: sinon.fake(),
    setNoDelay: sinon.fake(),
    on: sinon.fake(),
    write: sinon.fake()
  }
  const connect = sinon.fake.returns(socket)
  const log = sinon.fake()

  sinon.replace(connection === 'tcp' ? net : tls, 'connect', connect)

  if (echo) {
    sinon.replace(console, 'log', log)
  }

  const createWriteStream = api.createWriteStream

  const options = {
    appname: 'pino-papertrail',
    echo: echo,
    host: 'papertrailapp.com',
    port: '1234',
    connection: connection,
    'message-only': false
  }

  const writeStream = createWriteStream(options)
  t.ok(writeStream)
  t.ok(connect.called)

  t.ok(socket.on.calledTwice)
  const onErrorCall = socket.on.getCall(0)
  const onEndCall = socket.on.getCall(1)
  t.ok(onErrorCall.args[0] === 'error')
  t.ok(onEndCall.args[0] === 'end')

  // connection error (refuse)
  onErrorCall.lastArg(new Error('connect ECONNREFUSED'))
  clock.runAll()

  // should auto reconnect
  t.ok(connect.calledTwice)

  // should be able to write before a connection is made
  writeStream.write(`${messages.infoMessage}\n`, () => {
    t.ok(socket.write.notCalled)

    // accept connection, run callback
    connect.lastCall.lastArg()

    // should be able to write after a connection is made
    writeStream.write(`${messages.warnMessage}\n`, () => {
      t.ok(socket.write.calledTwice)
      t.ok(socket.write.getCall(0).args[0].toString().includes(messages.infoMessage))
      t.ok(socket.write.getCall(1).args[0].toString().includes(messages.warnMessage))

      // close connection
      onEndCall.lastArg()

      // should auto reconnect
      t.ok(connect.calledThrice)

      // accept connection, run callback
      connect.lastCall.lastArg()

      clock.uninstall()

      sinon.restore()

      t.end()
    })
  })
}

// sinon-using tests cannot be run concurrently
test('pino-papertrail api (tcp)', (t) => testApiTcp(t, 'tcp', false))
  .then(() => {
    test('pino-papertrail api (tls)', (t) => testApiTcp(t, 'tls', true))
  })


cli.js
#! /usr/bin/env node
'use strict'

const fs = require('fs')
const path = require('path')

const minimist = require('minimist')
const pumpify = require('pumpify')

const pinoPapertrail = require('./lib/pino-papertrail')
const pkg = require('./package.json')

const options = {
  alias: {
    version: 'v',
    help: 'h',
    echo: 'e',
    host: 'H',
    port: 'p',
    connection: 'c',
    appname: 'a',
    'message-only': 'm'
  },
  default: {
    appname: 'pino',
    echo: true,
    host: 'localhost',
    port: '1234',
    connection: 'udp',
    'message-only': false
  }
}

const argv = minimist(process.argv.slice(2), options)

if (argv.help) {
  console.log(fs.readFileSync(path.join(__dirname, './usage.txt'), 'utf8'))
  process.exit(0)
}
if (argv.version) {
  console.log(`${pkg.name} v${pkg.version}`)
  process.exit(0)
}

const parseJson = pinoPapertrail.parseJson()
const toSyslog = pinoPapertrail.toSyslog(argv)
const papertrail = pinoPapertrail.toPapertrail(argv)

function shutdown () {
  try {
    papertrail.close()
  } catch (e) {
    process.exit()
  }
}

process.on('SIGTERM', function () { shutdown() })

const pipeline = pumpify(parseJson, toSyslog, papertrail)
process.stdin.pipe(pipeline)


lib/pino-papertrail.js
'use strict'

const dgram = require('dgram')
const tls = require('tls')
const net = require('net')
const stream = require('stream')
const streamBuffers = require('stream-buffers')

const fastJsonParse = require('fast-json-parse')
const split2 = require('split2')
const glossy = require('glossy')
const through2 = require('through2')
const backoff = require('backoff')

const PINO_LEVELS = { trace: 10, debug: 20, info: 30, warn: 40, error: 50, fatal: 60 }
const SYSLOG_SEVERITIES = { emergency: 0, alert: 1, critical: 2, error: 3, warning: 4, notice: 5, info: 6, debug: 7 }

function _jsonParser (str) {
  const result = fastJsonParse(str)
  if (result.err) return
  return result.value
}

function _levelToSeverity (level) {
  if (level === PINO_LEVELS.trace || level === PINO_LEVELS.debug) { return SYSLOG_SEVERITIES.debug }
  if (level === PINO_LEVELS.info) { return SYSLOG_SEVERITIES.info }
  if (level === PINO_LEVELS.warn) { return SYSLOG_SEVERITIES.warning }
  if (level === PINO_LEVELS.error) { return SYSLOG_SEVERITIES.error }
  return SYSLOG_SEVERITIES.critical
}

function toPapertrailTcp (options) {
  var isConnected = false
  var socket = null

  // Create a socket with a retry strategy for any socket error
  function createSocket () {
    const backoffStrategy = options['backoff-strategy'] || new backoff.ExponentialStrategy()
    backoffStrategy.reset()
    const call = backoff.call((callback) => {
      socket = (options.connection === 'tls' ? tls : net).connect(options.port, options.host, () => {
        isConnected = true
        if (bufferStream.size() > 0) {
          const b = bufferStream.getContents()
          socket.write(b)
        }
      })
      socket.setKeepAlive(true)
      socket.setNoDelay(true)
      socket.on('error', (err) => {
        isConnected = false
        callback(err)
      })
      socket.on('end', () => {
        isConnected = false
        callback()
        createSocket().start()
      })
    }, () => {})
    call.setStrategy(backoffStrategy)
    call.on('backoff', (num, delay, err) => {
      console.log('Error: ' + err.message + '\nRetry in ' + delay + 'ms')
    })
    return call
  }
  createSocket().start()

  const bufferStream = new streamBuffers.WritableStreamBuffer()
  const writableStream = new stream.Writable({
    write (data, encoding, callback) {
      if (options.echo === true) { console.log(data.toString()) }

      if (isConnected) {
        socket.write(data + '\r\n', callback)
      } else {
        bufferStream.write(data + '\r\n', callback)
      }
    }
  })
  return writableStream
}

function toPapertrailUdp (options) {
  const socket = dgram.createSocket('udp4')
  const writableStream = new stream.Writable({
    write (data, encoding, callback) {
      if (options.echo === true) { console.log(data.toString()) }
      socket.send(data, 0, data.length, options.port, options.host, function (err) {
        if (err) {
          console.log('error', err.message)
        }
        callback()
      })
    }
  })
  return writableStream
}

module.exports.parseJson = function () {
  return split2(_jsonParser)
}

module.exports.toPapertrail = function (options) {
  if (options.connection === 'tls' || options.connection === 'tcp') {
    return toPapertrailTcp(options)
  } else {
    return toPapertrailUdp(options)
  }
}

module.exports.toSyslog = function (options) {
  const syslogProducer = new glossy.Produce()
  return through2.obj(function transport (data, enc, cb) {
    const msg = syslogProducer.produce({
      facility: options.facility || 'user',
      severity: _levelToSeverity(data.level),
      host: data.hostname,
      appName: options.appname,
      pid: data.pid,
      time: (data.time) ? new Date(data.time) : new Date(),
      message: options['message-only'] ? data.msg : JSON.stringify(data)
    })
    cb(null, msg)
  })
}
